src/main/java/org/tron/program/SolidityNode.java

```diff
@@ -23,157 +24,221 @@
+      for (int i = 0; i < 50; i++) {
+        new Thread(() -> getSyncBlock()).start();
+      }
+      new Thread(() -> getAdvBlock()).start();
+      new Thread(() -> pushBlock()).start();
+      new Thread(() -> processBlock()).start();
+      new Thread(() -> processTrx()).start();
+      logger.warn("Success to start solid node, lastSolidityBlockNum: {}.", lastSolidityBlockNum);
     } catch (Exception e) {
-      logger.error("Failed to create database grpc client {}", addr);
+      logger.error("Failed to start solid node, address: {}.", cfgArgs.getTrustNodeAddr());
       System.exit(0);
     }
   }
 
-  private void shutdownGrpcClient() {
-    if (databaseGrpcClient != null) {
-      databaseGrpcClient.shutdown();
+  private void getSyncBlock() {
+    long blockNum = getNextSyncBlockId();
+    while (syncFlag) {
+      try {
+        if (blockNum == 0) {
+          break;
+        }
+        if (blockMap.size() > 10000) {
+          sleep(1000);
+          continue;
+        }
+        blockMap.put(blockNum, databaseGrpcClient.getBlock(blockNum));
+        logger.info("Success to get sync block: {}.", blockNum);
+        blockNum = getNextSyncBlockId();
+      } catch (Exception e) {
+        logger.error("Failed to get sync block {}.", blockNum);
+        sleep(100);
+      }
     }
+    logger.warn("Get sync block thread {} exit.", Thread.currentThread().getName());
   }
 
-  private void syncLoop(Args args) {
-//    while (true) {
-//      try {
-//        initGrpcClient(args.getTrustNodeAddr());
-//        syncSolidityBlock();
-//        shutdownGrpcClient();
-//      } catch (Exception e) {
-//        logger.error("Error in sync solidity block " + e.getMessage(), e);
-//      }
-//      try {
-//        Thread.sleep(5000);
-//      } catch (InterruptedException e) {
-//        Thread.currentThread().interrupt();
-//        e.printStackTrace();
-//      }
-//    }
+  synchronized long getNextSyncBlockId() {
+
+    if (!syncFlag) {
+      return 0;
+    }
+
+    if (ID.get() < remoteLastSolidityBlockNum) {
+      return ID.incrementAndGet();
+    }
+
+    long lastNum = getLastSolidityBlockNum();
+    if (lastNum - remoteLastSolidityBlockNum > 50) {
+      remoteLastSolidityBlockNum = lastNum;
+      return ID.incrementAndGet();
+    }
+
+    logger.warn("Sync mode switch to adv, ID = {}, lastNum = {}, remoteLastSolidityBlockNum = {}",
+        ID.get(), lastNum, remoteLastSolidityBlockNum);
+
+    syncFlag = false;
+
+    return 0;
   }
 
-  private void syncSolidityBlock() throws BadBlockException {
-    DynamicProperties remoteDynamicProperties = databaseGrpcClient.getDynamicProperties();
-    long remoteLastSolidityBlockNum = remoteDynamicProperties.getLastSolidityBlockNum();
+  private void getAdvBlock() {
+    while (syncFlag) {
+      sleep(5000);
+    }
+    logger.warn("Get adv block thread start.");
+    long blockNum = ID.incrementAndGet();
     while (true) {

```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Add an end condition to this loop. ([squid:S2189](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2189))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1570&issues=AWYfmDct_XXc3YMrl43L&open=AWYfmDct_XXc3YMrl43L)

[ISSUE_KEY]: AWYfmDct_XXc3YMrl43L